gnome-initial-setup/gis-driver.c

```diff
@@ -288,6 +294,208 @@ gis_driver_get_mode (GisDriver *driver)
+                                   _("Failed to enter demo mode: %s"),
+                                   error->message);
+  gtk_dialog_run (GTK_DIALOG (dialog));
+  gtk_widget_destroy (dialog);
+  g_error_free (error);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ActUser, g_object_unref)
+
+static gboolean
+create_demo_user (GisDriver *driver)
+{
+  GisDriverPrivate *priv = gis_driver_get_instance_private (driver);
+  g_autoptr(ActUser) user;
+  GError *error = NULL;
+  const gchar *language;
+
+  error = NULL;
+  user = act_user_manager_create_user (act_user_manager_get_default (),
+                                       DEMO_ACCOUNT_USERNAME,
+                                       DEMO_ACCOUNT_FULLNAME,
+                                       ACT_USER_ACCOUNT_TYPE_STANDARD,
+                                       &error);
+  if (error)
+    {
+      handle_demo_mode_error (error);
+      return FALSE;
+    }
+
+  act_user_set_password_mode (user, ACT_USER_PASSWORD_MODE_NONE);
+  act_user_set_automatic_login (user, TRUE);
+
+  language = gis_driver_get_user_language (driver);
+
+  if (language)
+    act_user_set_language (user, language);
+
+  gis_driver_set_user_permissions (driver, user, NULL);
+  gis_update_login_keyring_password ("");
+
+  return TRUE;
+}
+
+#define GSD_POWER_SCHEMA "org.gnome.settings-daemon.plugins.power"
+#define GSD_SLEEP_INACTIVE_AC_TYPE "sleep-inactive-ac-type"
+#define GSD_SLEEP_INACTIVE_AC_TIMEOUT "sleep-inactive-ac-timeout"
+#define GSD_SLEEP_INACTIVE_BATTERY_TYPE "sleep-inactive-battery-type"
+#define GSD_SLEEP_INACTIVE_BATTERY_TIMEOUT "sleep-inactive-battery-timeout"
+
+#define ONE_MINUTE 60
+
+void
+configure_demo_mode_power_settings ()
+{
+  GSettings *power_settings = g_settings_new (GSD_POWER_SCHEMA);
+
+  g_settings_set_string (power_settings, GSD_SLEEP_INACTIVE_AC_TYPE, "logout");
+  g_settings_set_int (power_settings, GSD_SLEEP_INACTIVE_AC_TIMEOUT, ONE_MINUTE);
+  g_settings_set_string (power_settings, GSD_SLEEP_INACTIVE_BATTERY_TYPE, "logout");
+  g_settings_set_int (power_settings, GSD_SLEEP_INACTIVE_BATTERY_TIMEOUT, ONE_MINUTE);
+
+  g_object_unref (power_settings);
+}
+
+#define GS_SCHEMA "org.gnome.software"
+#define GS_ALLOW_UPDATES "allow-updates"
+
+void
+configure_demo_mode_disallow_app_center_updates ()
+{
+  GSettings *gnome_software_settings = g_settings_new (GS_SCHEMA);
+
+  g_settings_set_boolean (gnome_software_settings, GS_ALLOW_UPDATES, FALSE);
+
+  g_object_unref (gnome_software_settings);
+}
+
+#define GNOME_SHELL_SCHEMA "org.gnome.shell"
+#define GNOME_SHELL_FAVORITE_APPS "favorite-apps"
+
+#define GOOGLE_CHROME_DESKTOP_ID "google-chrome.desktop"
+#define CHROMIUM_BROWSER_DESKTOP_ID "chromium-browser.desktop"
+
+void
+replace_string_in_strv (GStrv       strv,
+                        const gchar *needle,
+                        const gchar *replacement)
+{
+  guint length = g_strv_length (strv);
+  guint i = 0;
+
+  for (; i < length; ++i)
```

Comment:
Nit: since `strv` is `NULL`-terminated it would be a bit more efficient to simply loop until `strv[i] != NULL`